File: mlp.py
import numpy as np


class LCG:
    def __init__(self, seed=42):
        self.a = 1664525
        self.c = 1013904223
        self.m = 2 ** 32
        self.state = seed

    def next(self):
        self.state = (self.a * self.state + self.c) % self.m
        return self.state

    def random(self):
        return self.next() / self.m


class MLP:
    def __init__(self, layers, learning_rate=0.01, seed=42):
        self.layers = layers
        self.learning_rate = learning_rate
        self.lcg = LCG(seed)
        self.weights = []
        self.biases = []
        self.initialize_parameters()

    def initialize_parameters(self):
        for i in range(len(self.layers) - 1):
            weight_matrix = np.zeros((self.layers[i], self.layers[i + 1]))
            bias_vector = np.zeros((1, self.layers[i + 1]))
            limit = np.sqrt(1 / self.layers[i])
            for j in range(self.layers[i]):
                for k in range(self.layers[i + 1]):
                    weight_matrix[j][k] = (self.lcg.random() * 2 - 1) * limit
            self.weights.append(weight_matrix)
            self.biases.append(bias_vector)

    def sigmoid(self, x):
        return 1 / (1 + np.exp(-x))

    def sigmoid_derivative(self, x):
        return x * (1 - x)

    def binary_cross_entropy(self, y_true, y_pred):
        epsilon = 1e-8
        y_pred = np.clip(y_pred, epsilon, 1 - epsilon)
        return - (y_true * np.log(y_pred) + (1 - y_true) * np.log(1 - y_pred))

    def binary_cross_entropy_derivative(self, y_true, y_pred):
        epsilon = 1e-8
        y_pred = np.clip(y_pred, epsilon, 1 - epsilon)
        return (y_pred - y_true) / (y_pred * (1 - y_pred))

    def forward(self, X):
        activations = [X]
        for w, b in zip(self.weights, self.biases):
            z = np.dot(activations[-1], w) + b
            a = self.sigmoid(z)
            activations.append(a)
        return activations

    def train(self, X, y, epochs):
        log_interval = epochs // 10 if epochs >= 10 else 1
        for epoch in range(1, epochs + 1):
            activations = self.forward(X)
            loss = np.mean(self.binary_cross_entropy(y, activations[-1]))

            # Backpropagation
            error = self.binary_cross_entropy_derivative(y, activations[-1])
            delta = error * self.sigmoid_derivative(activations[-1])

            for l in range(len(self.layers) - 2, -1, -1):
                a = activations[l]
                dw = np.dot(a.T, delta)
                db = np.sum(delta, axis=0, keepdims=True)
                if l != 0:
                    delta = np.dot(delta, self.weights[l].T) * self.sigmoid_derivative(activations[l])
                self.weights[l] -= self.learning_rate * dw
                self.biases[l] -= self.learning_rate * db

            if epoch % log_interval == 0 or epoch == 1:
                print(f"Epoch {epoch}/{epochs}, Loss: {loss:.6f}")

    def predict(self, X):
        activations = self.forward(X)
        return activations[-1]


# Usage Example
if __name__ == "__main__":
    # Define XOR dataset
    X = np.array([
        [0, 0],
        [0, 1],
        [1, 0],
        [1, 1]
    ])
    y = np.array([
        [0],
        [1],
        [1],
        [0]
    ])

    # Initialize MLP with [2,2,1], learning_rate=0.01, seed=42
    mlp = MLP(layers=[2, 2, 1], learning_rate=0.01, seed=42)

    # Train MLP
    epochs = 50000
    mlp.train(X, y, epochs)

    # Make predictions
    predictions = mlp.predict(X)
    predictions_binary = (predictions > 0.5).astype(int)

    print("\nPredictions after training:")
    for i, (input_sample, prediction) in enumerate(zip(X, predictions_binary)):
        print(f"Input: {input_sample}, Predicted Output: {prediction[0]}, True Output: {y[i][0]}")

File: mlp.cpp
// mlp.cpp
#include "mlp1.h"
#include <cmath>
#include <iostream>
#include <iomanip>

// Constructor
MLP::MLP(int input_size, int hidden_size, int output_size, double learning_rate, bool use_tanh_activation)
    : input_size(input_size), hidden_size(hidden_size),
      output_size(output_size), learning_rate(learning_rate),
      use_tanh(use_tanh_activation) {}

// Initialize parameters using LCG
void MLP::initialize_parameters(LCG &lcg) {
    // Initialize weights_input_hidden with Xavier initialization
    double limit_input_hidden = std::sqrt(1.0 / input_size);
    weights_input_hidden.resize(input_size, std::vector<double>(hidden_size, 0.0));
    for(int i = 0; i < input_size; ++i) {
        for(int j = 0; j < hidden_size; ++j) {
            double rand_uniform = lcg.random(); // [0,1)
            weights_input_hidden[i][j] = (rand_uniform * 2.0 - 1.0) * limit_input_hidden; // [-limit, limit)
        }
    }

    // Initialize bias_hidden to zeros
    bias_hidden.assign(hidden_size, 0.0);

    // Initialize weights_hidden_output with Xavier initialization
    double limit_hidden_output = std::sqrt(1.0 / hidden_size);
    weights_hidden_output.resize(hidden_size, std::vector<double>(output_size, 0.0));
    for(int i = 0; i < hidden_size; ++i) {
        for(int j = 0; j < output_size; ++j) {
            double rand_uniform = lcg.random(); // [0,1)
            weights_hidden_output[i][j] = (rand_uniform * 2.0 - 1.0) * limit_hidden_output; // [-limit, limit)
        }
    }

    // Initialize bias_output to zeros
    bias_output.assign(output_size, 0.0);
}

// Activation function: sigmoid
double MLP::activate(double x) {
    return 1.0 / (1.0 + std::exp(-x));
}

// Derivative of sigmoid
double MLP::activate_derivative(double activated_x) {
    return activated_x * (1.0 - activated_x);
}

// Binary Cross-Entropy Loss
double MLP::binary_cross_entropy(double y_true, double y_pred) {
    // Adding epsilon to prevent log(0)
    const double epsilon = 1e-8;
    y_pred = std::max(std::min(y_pred, 1.0 - epsilon), epsilon);
    return - (y_true * std::log(y_pred) + (1.0 - y_true) * std::log(1.0 - y_pred));
}

// Derivative of Binary Cross-Entropy Loss
double MLP::binary_cross_entropy_derivative(double y_true, double y_pred) {
    // Adding epsilon to prevent division by zero
    const double epsilon = 1e-8;
    y_pred = std::max(std::min(y_pred, 1.0 - epsilon), epsilon);
    return (y_pred - y_true) / (y_pred * (1.0 - y_pred));
}

// Forward Pass
void MLP::forward(const std::vector<double> &input,
                std::vector<double> &hidden_activations,
                std::vector<double> &output_activations) {

    // Calculate hidden layer activations
    hidden_activations.resize(hidden_size);
    for(int i = 0; i < hidden_size; ++i) {
        double activation = bias_hidden[i];
        for(int j = 0; j < input_size; ++j) {
            activation += input[j] * weights_input_hidden[j][i];
        }
        hidden_activations[i] = activate(activation);
    }

    // Calculate output layer activations
    output_activations.resize(output_size);
    for(int i = 0; i < output_size; ++i) {
        double activation = bias_output[i];
        for(int j = 0; j < hidden_size; ++j) {
            activation += hidden_activations[j] * weights_hidden_output[j][i];
        }
        output_activations[i] = activate(activation);
    }
}

// Backward Pass and Weights Update
void MLP::backward(const std::vector<double> &input,
                const std::vector<double> &hidden_activations,
                const std::vector<double> &output_activations,
                const std::vector<double> &y_true) {

    // Calculate output errors
    std::vector<double> output_errors(output_size, 0.0);
    for(int i = 0; i < output_size; ++i) {
        double derivative = activate_derivative(output_activations[i]);
        output_errors[i] = binary_cross_entropy_derivative(y_true[i], output_activations[i]) * derivative;
    }

    // Calculate hidden layer errors
    std::vector<double> hidden_errors(hidden_size, 0.0);
    for(int i = 0; i < hidden_size; ++i) {
        double derivative = activate_derivative(hidden_activations[i]);
        double error = 0.0;
        for(int j = 0; j < output_size; ++j) {
            error += output_errors[j] * weights_hidden_output[i][j];
        }
        hidden_errors[i] = error * derivative;
    }

    // Update weights_hidden_output and bias_output
    for(int i = 0; i < hidden_size; ++i) {
        for(int j = 0; j < output_size; ++j) {
            weights_hidden_output[i][j] -= learning_rate * output_errors[j] * hidden_activations[i];
        }
    }
    for(int i = 0; i < output_size; ++i) {
        bias_output[i] -= learning_rate * output_errors[i];
    }

    // Update weights_input_hidden and bias_hidden
    for(int i = 0; i < input_size; ++i) {
        for(int j = 0; j < hidden_size; ++j) {
            weights_input_hidden[i][j] -= learning_rate * hidden_errors[j] * input[i];
        }
    }
    for(int i = 0; i < hidden_size; ++i) {
        bias_hidden[i] -= learning_rate * hidden_errors[i];
    }
}

// Train the MLP
void MLP::train(const std::vector<std::vector<double>> &X,
               const std::vector<std::vector<double>> &Y,
               int epochs) {
    int log_interval = epochs / 10;
    if (log_interval == 0) log_interval = 1;

    for(int epoch = 1; epoch <= epochs; ++epoch) {
        double epoch_loss = 0.0;
        for(int i = 0; i < X.size(); ++i) {
            // Forward pass
            std::vector<double> hidden_activations;
            std::vector<double> output_activations;
            forward(X[i], hidden_activations, output_activations);

            // Calculate loss
            for(int j = 0; j < output_size; ++j) {
                epoch_loss += binary_cross_entropy(Y[i][j], output_activations[j]);
            }

            // Backward pass and update
            backward(X[i], hidden_activations, output_activations, Y[i]);
        }
        epoch_loss /= X.size();

        // Logging
        if(epoch % log_interval == 0 || epoch == 1) {
            std::cout << "Epoch " << epoch << "/" << epochs << ", Loss: " 
                      << std::fixed << std::setprecision(6) << epoch_loss << std::endl;
        }
    }
}

// Predict Output
std::vector<double> MLP::predict(const std::vector<double> &input) {
    std::vector<double> hidden_activations;
    std::vector<double> output_activations;
    forward(input, hidden_activations, output_activations);
    return output_activations;
}
